kvserver: clean up prepareSnapApply

pav-kv opened the pull request:
This PR introduces a builder type which prepares a set of writes when applying a snapshot. This removes extensive lists of parameters dragged across function calls, and wraps repetitive SST creation code into a helper method.

It also fixes a TODO in `TestPrepareSnapApply`: all the writes are now printed in the original order, thanks to the `storage.Writer` abstraction instead of direct SST generation code.

The new `storage.Writer`-based interface is more flexible, and e.g. allows eliminating the `ConvertFilesToBatchAndCommit` [quirk](https://github.com/cockroachdb/cockroach/blob/255f60e14a03060b9b2ca81e38f7188c33173a7a/pkg/kv/kvserver/replica_raftstorage.go#L640) as well as the "cleared spans" tracking. We can know when the snapshot is ["simple/small"](https://github.com/cockroachdb/cockroach/blob/255f60e14a03060b9b2ca81e38f7188c33173a7a/pkg/kv/kvserver/replica_raftstorage.go#L624-L627) before `MultiSSTWriter` starts writing files (e.g. from the handshake), so we can generate a batch directly and skip the files writing phase; this batch can then be extended with `prepareSnapApply` data and committed. This would make small snapshot ingestions faster.

Epic: CRDB-49111

pav-kv committed:
kvserver: make snapshot builder type

Epic: none
Release note: none

pav-kv committed:
kvserver: make helper for SST creation

Epic: none
Release note: none

pav-kv committed:
kvserver: make rewriteRaftState a method

Epic: none
Release note: none

pav-kv committed:
kvserver: make clearSubsumedReplicaDiskData a method

Epic: none
Release note: none

pav-kv committed:
kvserver: store cleared spans in the builder struct

Epic: none
Release note: none

pav-kv committed:
kvserver: wrap snapshot SST creations

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/149499)
<!-- Reviewable:end -->

pav-kv committed:
kvserver: move SST wrapper to SSTStorageScratch

This creates a more concise abstraction in snapWriteBuilder, and allows
printing the writes in the right order in tests.

Epic: none
Release note: none

pav-kv committed:
spanrecv: squash WriteSST method

Epic: none
Release note: none

pav-kv commented on pkg/kv/kvserver/kvstorage/snaprecv/sst_snapshot_storage.go:
```diff
@@ -138,21 +141,34 @@ func (s *SSTSnapshotStorageScratch) NewFile(
 	}
-	if len(data) == 0 {
+
+	// TODO(itsbilal): Write to SST directly rather than buffer in a MemObject.
```
This is a questionable TODO. We could move `s.NewFile()` from below right here, and pass `f` into `MakeIngestionSSTWriter()`, but it would run every key write through this path (unless the SST writer has some buffer of its own):

https://github.com/cockroachdb/cockroach/blob/63545b4d7dfd9ac76f62b4b0ddd580b000980309/pkg/kv/kvserver/kvstorage/snaprecv/sst_snapshot_storage.go#L229-L231

Leaving this TODO for now.

pav-kv commented on pkg/kv/kvserver/testdata/TestPrepareSnapApply.txt:
```diff
@@ -1,32 +1,32 @@
-set: /Local/RangeID/123/u/RaftReplicaID/0,0 -> replica_id:4 
-set: /Local/RangeID/123/u/RaftTruncatedState/0,0 -> index:100 term:20 
-rangedel: /Local/RangeID/123/{u""-v""}
+Delete Range: [0,0 /Local/RangeID/123/u"" (0x0169f67b7500): , 0,0 /Local/RangeID/123/v"" (0x0169f67b7600): )
+Put: 0,0 /Local/RangeID/123/u/RaftHardState (0x0169f67b757266746800): term:20 vote:0 commit:100 lead:0 lead_epoch:0 
+Put: 0,0 /Local/RangeID/123/u/RaftReplicaID (0x0169f67b757266747200): replica_id:4 
+Put: 0,0 /Local/RangeID/123/u/RaftTruncatedState (0x0169f67b757266747400): index:100 term:20 
```
I kinda like the previous, more concise format more. Might fix it in the future.

pav-kv committed:
kvserver: pass context into write functor

Epic: none
Release note: none

pav-kv committed:
kvserver: rm unnecessary parameter

Epic: none
Release note: none

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 8 of 8 files at r10, 4 of 4 files at r11, 1 of 1 files at r12, 8 of 8 files at r14, 8 of 8 files at r15, 1 of 1 files at r16, 3 of 3 files at r17, 1 of 1 files at r18, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149499#-:-OUL16QJ2VSatnNKEvOX:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @tbg)



<!-- Sent from Reviewable.io -->
